addOns/spider/src/main/java/org/zaproxy/addon/spider/DomainAlwaysInScopeMatcher.java

```diff
@@ -0,0 +1,137 @@
+
+/**
+ * Class that contains rules to checks if a domain is in scope.
+ *
+ * <p>It supports both plain text and regular expression checks.
+ *
+ * @see #matches(String)
+ */
+public class DomainAlwaysInScopeMatcher extends Enableable {
+
+    private final Pattern pattern;
+    private final String domain;
+    private final boolean regex;
+
+    public DomainAlwaysInScopeMatcher(Pattern pattern) {
+        super(true);
+
+        if (pattern == null) {
+            throw new IllegalArgumentException("Parameter pattern must not be null.");
+        }
+
+        this.pattern = pattern;
+        this.regex = true;
+        this.domain = null;
+    }
+
+    public DomainAlwaysInScopeMatcher(String domain) {
+        super(true);
+
+        if (domain == null || domain.isEmpty()) {
+            throw new IllegalArgumentException("Parameter domain must not be null or empty.");
+        }
+
+        this.domain = domain;
+        this.regex = false;
+        this.pattern = null;
+    }
+
+    public DomainAlwaysInScopeMatcher(DomainAlwaysInScopeMatcher other) {
+        super(other.isEnabled());
+
+        this.domain = other.domain;
+        this.regex = other.regex;
+        this.pattern = other.pattern;
+    }
+
+    public String getValue() {
+        if (isRegex()) {
+            return pattern.pattern();
+        }
+
+        return domain;
+    }
+
+    public boolean isRegex() {
+        return regex;
+    }
+
+    /**
+     * Tells whether or not the given domain is considered always in scope.
+     *
+     * @param domain the domain that will be checked
+     * @return {@code true} if the domain is considered always in scope, {@code false} otherwise.
+     */
+    public boolean matches(String domain) {
+        if (pattern != null) {
+            return pattern.matcher(domain).matches();
+        }
+
+        return this.domain.equals(domain);
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = super.hashCode();
+        result = prime * result + ((domain == null) ? 0 : domain.hashCode());
+        result = prime * result + ((pattern == null) ? 0 : pattern.pattern().hashCode());
+        return result;
+    }
+
+    @Override
+    public boolean equals(Object object) {
+        if (this == object) {
+            return true;
+        }
+        if (!super.equals(object)) {
+            return false;
+        }
+        if (getClass() != object.getClass()) {
+            return false;
+        }
+        DomainAlwaysInScopeMatcher other = (DomainAlwaysInScopeMatcher) object;
+        if (domain == null) {
+            if (other.domain != null) {
+                return false;
+            }
+        } else if (!domain.equals(other.domain)) {
+            return false;
+        }
+        if (pattern == null) {
+            if (other.pattern != null) {
+                return false;
+            }
+        } else if (!pattern.pattern().equals(other.pattern.pattern())) {
```

Comment:
LGTM claims pattern can be null here. However if I add `pattern != null &&` SonarLint says it always evaluates to true.

Reply:
Needs `other.pattern == null ||`.

Reply:
okay